### PR TITLE
ta: pkcs11: Some fixes and improvements in C_DeriveKey() flow to make it more generic and reusable for C_UnwrapKey()

### PR DESCRIPTION
**Pull request by ruchi393**

### CODE REVIEW

**Comment by ruchi393** [comment]
Force pushed after fixing Shippable errors

**Comment by etienne-lms** [comment]
azure pipes found:
```
2021-03-04T04:25:38.3659835Z ta/pkcs11/src/processing.c: In function 'entry_processing_key':
2021-03-04T04:25:38.3661549Z ta/pkcs11/src/processing.c:579:50: error: passing argument 3 of 'derive_key_by_symm_enc' from incompatible pointer type [-Werror=incompatible-pointer-types]
2021-03-04T04:25:38.3662366Z    rc = derive_key_by_symm_enc(session, &out_buf, &out_size);
2021-03-04T04:25:38.3662957Z                                                   ^
2021-03-04T04:25:38.3663511Z In file included from ta/pkcs11/src/processing.c:19:0:
2021-03-04T04:25:38.3664542Z ta/pkcs11/src/processing.h:62:16: note: expected 'size_t * {aka long unsigned int *}' but argument is of type 'uint32_t * {aka unsigned int *}'
2021-03-04T04:25:38.3665298Z  enum pkcs11_rc derive_key_by_symm_enc(struct pkcs11_session *session,
2021-03-04T04:25:38.3665846Z                 ^~~~~~~~~~~~~~~~~~~~~~
2021-03-04T04:25:38.3666345Z cc1: all warnings being treated as errors

**Comment by ruchi393** [comment]
> derive_key_by_symm_enc

My bad. I fixed the issue in C file. Forgot to change the prototype in header file. I have forced pushed the commits.

File: ta/pkcs11/src/pkcs11_attributes.c [hunk 1]
```diff
@@ -1603,3 +1603,53 @@ enum pkcs11_rc check_attrs_against_modification(struct pkcs11_session *session,
...
+			    size_t key_size)
+{
+	switch (get_class(*head)) {
+	case PKCS11_CKO_SECRET_KEY:
+		return set_secret_key_data(head, data, key_size);
+	default:
+		return PKCS11_CKR_GENERAL_ERROR;
+	}
+
+	return PKCS11_CKR_OK;
```
Comment (etienne-lms) [inline ta/pkcs11/src/pkcs11_attributes.c:1654]: can remove these 2 lines

File: ta/pkcs11/src/pkcs11_attributes.c [hunk 2]
```diff
@@ -1603,3 +1603,53 @@ enum pkcs11_rc check_attrs_against_modification(struct pkcs11_session *session,
...
+	uint32_t key_length = 0;
+	enum pkcs11_rc rc = PKCS11_CKR_GENERAL_ERROR;
+
+	/* Get key size if present in template */
+	rc = get_attribute(*head, PKCS11_CKA_VALUE_LEN, &key_length, &size);
+	if (rc && rc != PKCS11_RV_NOT_FOUND)
+		return rc;
+
+	if (key_length) {
+		/* Key received is smaller than the required size */
```
Comment (etienne-lms) [inline ta/pkcs11/src/pkcs11_attributes.c:1620]: can remove this inline comment

File: ta/pkcs11/src/processing.c [hunk 3]
```diff
@@ -579,15 +574,22 @@ enum pkcs11_rc entry_derive_key(struct pkcs11_client *client,
...
 
-		rc = derive_key_by_symm_enc(session, &head);
+		session->processing->mecha_type = proc_params->id;
+
+		rc = derive_key_by_symm_enc(session, &out_buf, &out_size);
 		if (rc)
 			goto out;
-		break;
-	default:
+	} else {
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing.c:582]: add `goto out;` below

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for
* ta: pkcs11: Correct error returned when processing mechanisms
* ta: pkcs11: Remove check from entry_derive_key()
* ta: pkcs11: Rename entry_derive_key() to make it more generic
(note for this one, in commit log: s/underiveping/deriving/)

File: ta/pkcs11/src/processing.c (hunk 3 above)
Comment (ruchi393) [reply ta/pkcs11/src/processing.c:582]: Thanks for catching this. Fixing it.

File: ta/pkcs11/src/pkcs11_attributes.c (hunk 2 above)
Comment (ruchi393) [reply ta/pkcs11/src/pkcs11_attributes.c]: Removed

File: ta/pkcs11/src/pkcs11_attributes.c (hunk 1 above)
Comment (ruchi393) [reply ta/pkcs11/src/pkcs11_attributes.c]: Removed

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

**Comment by ruchi393** [comment]
Thanks @etienne-lms for reviewing the patches.

Patches Squashed, commit message updated s/underiveping/unwrapping. Tags applied.

**Review by etienne-lms — APPROVED** [review]

**Pull request merged**